### ISSUE TITLE
Issue with Bytes32 implementation

### ISSUE THREAD

**Issue by jjstew50** [ethereum/go-ethereum#26532]
https://github.com/ethereum/go-ethereum/blob/2b44ef5f93cc7479a77890917a29684b56e9167a/signer/core/apitypes/types.go#L387

This if statement makes it so all types if there is a type will fall into this statement. I am trying to encode a bytes32 array and I think it should go to the encodePrimitive function instead but the code is showing this is not possible. Can someone help me with this encoding?

```
signerData := signer.TypedData{
Types: signer.Types{
	"SignedMessage": []signer.Type{
		{Name: "key", Type: "bytes32"},
		{Name: "sender", Type: "address"},
		{Name: "amount", Type: "uint256"},
	},
	"EIP712Domain": []signer.Type{
		{Name: "name", Type: "string"},
		{Name: "version", Type: "string"},
		{Name: "chainId", Type: "uint256"},
	},
},
// PrimaryType: "SignedMessage",
Domain: signer.TypedDataDomain{
	Name:    "Neptune Bridge",
	Version: "0.0.1",
	ChainId: chainId,
},
Message: signer.TypedDataMessage{
	"key":    key,
	"sender": account,
	"amount": bridgedAmount,
},
}

typedDataHash, err := signerData.HashStruct("SignedMessage", signerData.Message)
if err != nil {
log.Fatal(err)
}
fmt.Println("\nTypedDataHash: ",typedDataHash)
```

This code snippet will return a different hash than in solidity:
```
        bytes32 hashToVerify = keccak256(
            abi.encode(keccak256("SignedMessage(bytes32 key, address sender, uint256 amount)"), key, sender, amount)
        );

        return hashToVerify;
```

I was able to verify my domainSeparatorHash were the same in solidity and go. But the message is now hashing the same between the two. Can someone help me with this? Thanks!

**Comment by holiman** [comment]
Could you please provide the output that you obtain for the first snippet, and also solidity?

**Comment by jjstew50** [comment]
Figured out my issue in solidity I was using:
`        bytes32 hashToVerify = keccak256(
            abi.encode(keccak256("SignedMessage(bytes32 key, address sender,uint256 amount)"),nonce,sender,bridgedAmount)
        );`

When it shouldve been:
``        bytes32 hashToVerify = keccak256(
            abi.encode(keccak256("SignedMessage(bytes32 key,address sender,uint256 amount)"),nonce,sender,bridgedAmount)
        );`

I had added in a space between `key,` and  `address` because this was a string input it encoded differently that the ethereum-go code. This was a pain in the butt to figure out the issue but the issue was in my solidity not go code. For anyone else out there, check that you don't have spaces if you are hardcoding values in solidity.
